### PR TITLE
(PDB-1738) (PDB-1739) Drop jdbc.deprecated

### PR DESCRIPTION
**Pull request by rbrw**
```
(PDB-1738) (PDB-1739) Drop jdbc.deprecated

Migrate all code to the newer clojure.jdbc API.

Note that the newer jdbc API handles some database operations like
transactions by storing bookkeeping information in nested database
references.  For example, (with-db-transaction ...) creates a new
database connection map with an incremented level, and the library uses
that level to determine where the "top" is, i.e. to decide when to
commit.

To accommodate that approach (without threading a db argument through
all of the existing code), add a jdbc/*db* variable to refer to the
"current" database, and make sure to bind appropriately it in
puppetdb.jdbc wrappers for the relevant clojure.jdbc calls.

Use (jdbc/db) to refer to *db* when invoking the new jdbc API's
functions.

Use jdbc/ consistently to refer to puppetdb.jdbc in altered files.

While updating migrations, combine adjacent db-do-commands into one
broader do-commands.
```

### CODE REVIEW

File: src/puppetlabs/puppetdb/cli/services.clj [hunk 1]
```diff
@@ -54,7 +53,8 @@
             [puppetlabs.puppetdb.command.dlo :as dlo]
             [puppetlabs.puppetdb.config :as conf]
             [puppetlabs.puppetdb.http.server :as server]
-            [puppetlabs.puppetdb.jdbc :as pl-jdbc :refer [with-transacted-connection]]
+            [puppetlabs.puppetdb.jdbc :as jdbc
+             :refer [with-transacted-connection]]
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/cli/services.clj]: might as well move calls to `with-transacted-connection` under the `jdbc` ns as well

File: src/puppetlabs/puppetdb/jdbc.clj [hunk 2]
```diff
@@ -63,12 +79,12 @@
...
    :post [(or (zero? limit) (<= (count %) limit))]}
   (let [sql-query-and-params (if (string? query) [query] query)]
-    (sql/with-query-results result-set
-      sql-query-and-params
-      (let [limited-result-set (limit-result-set! limit result-set)]
-        (-> limited-result-set
-            convert-result-arrays
-            vec)))))
+    (sql/db-query-with-resultset
+     (jdbc/db)
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/jdbc.clj]: I don't think you need `jdbc/` here

File: src/puppetlabs/puppetdb/jdbc.clj [hunk 3]
```diff
@@ -2,17 +2,33 @@
...
   (:import (com.jolbox.bonecp BoneCPDataSource BoneCPConfig)
            (java.util.concurrent TimeUnit))
-  (:require [clojure.java.jdbc.deprecated :as sql]
+  (:require [clojure.java.jdbc :as sql]
             [clojure.string :as string]
             [clojure.tools.logging :as log]
             [puppetlabs.kitchensink.core :as kitchensink]
             [clojure.string :as str]
             [puppetlabs.puppetdb.time :as pl-time]
+            [puppetlabs.puppetdb.jdbc :as jdbc]
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/jdbc.clj]: Think this is unnecessary, since this is the ns we're in

File: src/puppetlabs/puppetdb/query_eng.clj [hunk 4]
```diff
@@ -205,6 +205,6 @@
                               INNER JOIN factsets
                               ON factsets.certname = certnames.certname
                               WHERE certnames.certname=?")]
-    (sql/with-query-results result-set
-      [check-sql id]
-      (pos? (count result-set)))))
+    (sql/db-query-with-resultset (jdbc/db)
+                                 [check-sql id]
+                                 #(boolean (seq (sql/result-set-seq %))))))
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/query_eng.clj]: I think this should probably be a `comp`

File: test/puppetlabs/puppetdb/scf/migrate_test.clj [hunk 5]
```diff
@@ -217,22 +222,21 @@
...
-          :value_type_id 1
-          :value_hash (hash/generic-identity-hash 1)
-          :value_integer 1}))
+      (let [[p1 p2]
+            (map :id (sql/insert! (jdbc/db) :fact_paths
+                                  {:value_type_id 0
+                                   :path "orphan" :name "orphan" :depth 0}
+                                  {:value_type_id 1
+                                   :path "orphan" :name "orphan" :depth 0}))]
+        (sql/insert! (jdbc/db) :fact_values
```
Comment (ajroetker) [inline test/puppetlabs/puppetdb/scf/migrate_test.clj]: can we just make our own `jdbc/insert!` which does this? Similar for `db-do-commands`? I think that's what @wkalt had in mind for the dynamic var.

Comment (wkalt) [inline test/puppetlabs/puppetdb/scf/migrate_test.clj]: +1 to that

Comment (rbrw) [inline test/puppetlabs/puppetdb/scf/migrate_test.clj]: Let's chat about this - I've gotten the impression there are varying opinions (though none categorical), and I started these changes accordingly, but as a practical matter I suspect we might be better off if I rework this (perhaps more in the direction you suggest).

**Comment by rbrw** [comment]
@pljenkinsro retest this please

**Comment by rbrw** [comment]
@pljenkinsro retest this please

**Comment by wkalt** [comment]
test this please

**Comment by wkalt** [comment]
@puppetlabs-jenkins test this please

**Comment by wkalt** [comment]
this needs a rebase

**Comment by rbrw** [comment]
@puppetlabs-jenkins retest this please...

**Comment by wkalt** [comment]
@puppetlabs-jenkins retest this please

**Comment by senior** [comment]
@rbrw travis is failing right now, looks like migrate-test is hanging?

**Comment by senior** [comment]
I think there's a memory leak here. Travis doesn't say OutOfMemoryError, but I see it locally. For me it ran out of memory at storage-test, but only when I ran the whole test suite. If I run just that test it's fine. Also I only saw this on HyperSQL.

File: src/puppetlabs/puppetdb/scf/storage.clj [hunk 6]
```diff
@@ -925,12 +942,13 @@
...
 (defn realize-records!
-  "Inserts the records (maps) into the named database and returns them
+  "Inserts the records (maps) into the named table and returns them
   with their new :id values."
-  [database records]
-  (map #(assoc %2 :id %1)
-       (map :id (apply sql/insert-records database records))
-       records))
+  [table records]
+  (when (seq records)
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/scf/storage.clj]: did something change to require this?

File: test/puppetlabs/puppetdb/fixtures.clj [hunk 7]
```diff
@@ -17,20 +17,20 @@
             [puppetlabs.puppetdb.scf.migrate :refer [migrate!]]
             [puppetlabs.puppetdb.middleware :as mid]))
 
-(def ^:dynamic *db* nil)
+(def ^:dynamic *db* nil) ;; FIXME: perhaps this should be *db-spec* (cf db.clj)?
```
Comment (wkalt) [inline test/puppetlabs/puppetdb/fixtures.clj]: *db* is fine with me

File: test/puppetlabs/puppetdb/jdbc_test.clj [hunk 8]
```diff
@@ -1,36 +1,46 @@
...
+                 (subject/query-to-vec
+                  "SELECT value FROM test WHERE key='foo'"))))))
+    (let [read-pool (-> (assoc *db-spec* :read-only? true)
+                        fixt/defaulted-read-db-config
+                        subject/pooled-datasource)]
+      (with-open [_ (:datasource read-pool)]
+        (subject/with-transacted-connection read-pool
+          (let [msg (try
+                      (insert-map {"bar" 1})
+                      ""
```
Comment (wkalt) [inline test/puppetlabs/puppetdb/jdbc_test.clj]: what does this do?

Comment (wkalt) [inline test/puppetlabs/puppetdb/jdbc_test.clj]: makes failures prettier I guess

File: src/puppetlabs/puppetdb/scf/storage.clj [hunk 9]
```diff
@@ -1230,8 +1243,7 @@
                :where [:= :certname certname]
                :order-by [[:producer_timestamp :desc]]
                :limit 1}]
-    (sql/with-query-results result-set (hcore/format query)
-      (:producer_timestamp (first result-set)))))
+    (:producer_timestamp (first (jdbc/query (hcore/format query))))))
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/scf/storage.clj]: we're using query here and in a couple other spots while it seems like we're using query-with-resultset in most other similar places. I don't mind, but is there any particular reason?

File: src/puppetlabs/puppetdb/jdbc.clj [hunk 10]
```diff
@@ -191,23 +262,26 @@
      false)))
 
 (pls/defn-validated retry-sql*
-  "Executes f. If an exception is thrown, will retry. At most n retries
-   are done. If still some retryable error state is thrown it is bubbled upwards
-   in the call chain."
-  [remaining :- s/Int
+  "Inovkes (f) up to n times, retrying only if a transient connection
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/jdbc.clj]: invokes

**Comment by wkalt** [comment]
managed to get all the way through this. Generally +1, just a few clarifying comments.

File: src/puppetlabs/puppetdb/scf/storage.clj (hunk 6 above)
Comment (rbrw) [inline src/puppetlabs/puppetdb/scf/storage.clj]: Yes - I believe insert! chokes when given no arguments, and insert-records didn't.

File: test/puppetlabs/puppetdb/fixtures.clj (hunk 7 above)
Comment (rbrw) [inline test/puppetlabs/puppetdb/fixtures.clj]: OK, the context was that I think the new API may behave differently in some cases when given a an actual DB connection as compared to a config (i.e. spec).  For example, if I recall correctly, if you give with-db-connection a db, it won't open a new connection, it'll just use the one you give it (and close it), whereas "with-connection" (the old one) would create a new connection.  I believe that's what made me wonder.

File: test/puppetlabs/puppetdb/jdbc_test.clj (hunk 8 above)
Comment (rbrw) [inline test/puppetlabs/puppetdb/jdbc_test.clj]: One change here was to use with-open so that we always close.  Another was to separate\ the read and write pool tests, since they didn't share the connections, and finally, the "read-only" string moved.  It's no longer in the first exception in the "get cause" list, so I added the full-sql-exception-msg function to flatten the text (which might be what you were really asking about...).

File: src/puppetlabs/puppetdb/scf/storage.clj (hunk 9 above)
Comment (rbrw) [inline src/puppetlabs/puppetdb/scf/storage.clj]: I tried to make sure I only used query if I knew the query had a (small) result limit, i.e. "limit 1", since I believe query is not lazy; otherwise, I used db-query-with-resultset.

**Comment by rbrw** [comment]
@wkalt thanks for the slog - if my explanations sound plausible, then I'll push the typo fix, and perhaps we'll be set.

File: test/puppetlabs/puppetdb/jdbc_test.clj (hunk 8 above)
Comment (wkalt) [inline test/puppetlabs/puppetdb/jdbc_test.clj]: heh, I actually just meant the empty string on line 40. My guess is that when the test fails, that allows it to register as a failure rather than an error in re-find.

Comment (rbrw) [inline test/puppetlabs/puppetdb/jdbc_test.clj]: Yes, I think that's right.

File: src/puppetlabs/puppetdb/scf/storage.clj [hunk 11]
```diff
@@ -262,24 +261,26 @@
...
   "Returns the id (primary key) from `table` that contain `row-map` values"
   [table :- s/Keyword
    row-map :- {s/Keyword s/Any}]
   (let [cols (keys row-map)
-        where-clause (str "where " (str/join " " (map (fn [col] (str (name col) "=?") ) cols)))]
-    (sql/with-query-results rs (apply vector (format "select id from %s %s" (name table) where-clause) (map row-map cols))
-      (:id (first rs)))))
+        q (format "select id from %s where %s"
+                  (name table)
+                  (str/join " " (map #(str (name %) "=?") cols)))]
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/scf/storage.clj]: Not part of this patch, but this seems like a good candidate for honeysql.

File: src/puppetlabs/puppetdb/scf/migrate.clj [hunk 12]
```diff
@@ -993,133 +989,139 @@
...
+              SELECT id, " (munge-hash "value_hash") ", value_type_id,
+                     value_integer, value_float, value_string, value_boolean,
+                     value_json
+                FROM fact_values")
 
-    (sql/create-table :resource_params_cache_transform
-                      ["resource" hash-type "NOT NULL"]
-                      ["parameters" "TEXT"])
+      (sql/create-table-ddl :resource_params_cache_transform
+                             ["resource" hash-type "NOT NULL"]
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/scf/migrate.clj]: alignment

File: src/puppetlabs/puppetdb/scf/storage.clj [hunk 13]
```diff
@@ -199,15 +199,14 @@
   "Returns a boolean indicating whether or not the given certname exists in the db"
   [certname]
   {:pre [certname]}
-  (sql/with-query-results result-set
-    ["SELECT 1 FROM certnames WHERE certname=? LIMIT 1" certname]
-    (pos? (count result-set))))
+  (not (empty? (jdbc/query ["SELECT 1 FROM certnames WHERE certname=? LIMIT 1"
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/scf/storage.clj]: `not ( empty?` -> `some?`

**Comment by ajroetker** [comment]
Lgtm other than small comments, will there need to be a corresponding PR for PE as well?

File: src/puppetlabs/puppetdb/scf/storage.clj (hunk 13 above)
Comment (rbrw) [inline src/puppetlabs/puppetdb/scf/storage.clj]: @ajroetker hmm, I believe query returns a (possibly empty) vector, not nil, so we may need empty? here.

Comment (rbrw) [inline src/puppetlabs/puppetdb/scf/storage.clj]: @ajroetker I take that back, the default for as-arrays? is false, so it calls doall on the result, not vec.

Comment (rbrw) [inline src/puppetlabs/puppetdb/scf/storage.clj]: ...in any case, we'd still need (some? (seq (query ...)))

**Comment by rbrw** [comment]
@ajroetker looks like we're fine there.

**Comment by ajroetker** [comment]
Well I g2g on this once the acceptance tests pass!
